UnitTest/tests/test_SDUT-devices.txt

```diff
@@ -0,0 +1,200 @@
+															"model"      => "Techmar"
+														},
+		"Tedsen_SKX2xx_1F10110" =>  {	"DEF"          => "Tedsen_SKX2xx 1F10110",
+																	"rmsg"         => "MU;P0=-2009;P1=2076;P2=-259;P3=329;P4=-15751;P5=1216;D=012121212303012303034121212301212303012121212303012303034121212301212303012121212303012303034121212305;CP=3;R=37;",
+																	"state"        => "Button_1",
+																	"tristateCode" => "1F10110F0",
+																	"model"        => "Tedsen_SKX2xx"
+																},
+		"Tedsen_SKX6xx_1F10FF0" =>  {	"DEF"          => "Tedsen_SKX6xx 1F10FF0",
+																	"rmsg"         => "MU;P0=-32001;P1=2072;P2=-260;P3=326;P4=-2015;P5=-15769;D=01212123412123434123412343434123412151212123412123434123412343434123412151212123412123434123412343434123412151212123412123434123412343434123412151212123412123434123412343434123412151212123412123434123412343434123412151212123412123434123412343434123412151;CP=3;R=37;O;",
+																	"state"        => "Button_6",
+																	"tristateCode" => "1F10FF0F1",
+																	"model"        => "Tedsen_SKX6xx"
+																},
+		"Unitec_47031_95" =>  {	"DEF"   => "Unitec_47031 95",
+														"rmsg"  => "MU;P0=309;P1=636;P2=-690;P3=-363;P4=-10027;D=012031203120402031312031203120312031204020313120312031203120312040203131203120312031203120402031312031203120312031204020313120312031203120312040203131203120312031203120402031312031203120312031204020313120312031203120312040203131203120312030;CP=0;O;",
+														"state" => "1",
+														"Zone"  => "010 (off|on|off) - Zone 3",
+														"ToDo"  => "added to JSON and recheck",
+														"model" => "Unitec_47031"
+													},
+	);
+	
+	for my $dev (sort keys %devices) {
+		subtest "Checking $dev" => sub {
+			my $plan = 0;
+
+			for my $key (keys %{$devices{$dev}}) {
+				#note($key);
+				$plan++ if ($key ne "DEF" && $key ne "rmsg" && $key ne "ToDo" && $key ne "model");
+			}
+			
+			plan => $plan;
+
+			if(defined($defs{$dev}) && $defs{$dev}{TYPE} eq "SD_UT" && $defs{$dev}{NAME} eq $dev) {
+				note("device is already defined on system");
+				CommandSetReading(undef, "$dev state missing");
+				#CommandDelete(undef,$dev);
+			} else {
+				note("device must defined");
+				CommandDefMod(undef,"-temporary $dev SD_UT ".$devices{$dev}{DEF});			
+			}
+
+			my %signal_parts = SIGNALduino_Split_Message($devices{$dev}{rmsg},$targetHash->{NAME});
+			SIGNALduino_Parse($targetHash, $targetHash, $targetHash->{NAME}, "\002".$devices{$dev}{rmsg}."\003");
+			
+			## all readings, if ne DEF,rmsg,ToDo ##
+			for my $key (keys %{$devices{$dev}}) {
+				#note($key);
+				is(ReadingsVal($dev ,$key,"0"),$devices{$dev}{$key},"check reading $key") if ($key ne "DEF" && $key ne "rmsg" && $key ne "ToDo" && $key ne "model");
+			}
+
+			TODO: {
+				# read from internal STATE, if FHEM restart -> defined, after dispatch must changed !!
+				is(InternalVal($dev ,"STATE","0"),$devices{$dev}{state},"check internal state");
```

Comment:
Wenn ich das korrekt überblicke, dann wird der Wert {state} gegen das Reading state und auch gegen das INTERNAL STATE verglichen.

Korrekt?

Reply:
Das ist richtig.

Grund, wenn ich den Test lokal laufen lasse und das Device ist schon angelegt, modifiziere ich den state, da dieser den alten richtigen Wert haben könnte und somit ich ein INOrdnung erhalte obwohl die Message nicht ankam. Die doppelte Prüfung ist ebenso, weil bei einem FHEM restart das Internal STATE auf DEFINED steht und im REading state der alte Wert. Beides muss stimmen, somit kam die Nachricht auch wirklich an.

Reply:
Ja verstehe.
Ich kenne das. Vor Beginn eines Tests sollte man die "Standard" Werte definieren, sonst kann der Test verfälscht werden.
Mit setReading könntest Du vorher die Readings, welche getestet werden sollen z.B. mit NULL füllen.

Reply:
Das mache ich ja schon mit dem Reading state. Somit habe ich auch den Fehler erkannt, das bei einem Device derzeit die Nachricht ankommt. Alle anderen Readings sehe ich erstmal 2. Ranging weil das Ankommen des states für mich zeigt, es kam durch.

Jedes Reading, das wäre dann direkt auf den Code getestet noch tiefer, das kann später mal vielleicht. Wir müssen erstmal abfangen, das auch wirklcih die ca. 15 Devices in dem Modul weiterhin immer funktionieren.

Reply:
Okay, hatte ich übersehen.

Ich finde dein Framework wirklich top.
Das ist doch schon jetzt so universell um es auch für andere Tests verwenden zu können.

Reply:
Kleine Aufgabe mal für dich. Prüfe mal bitte, ob es machbar ist mit kleinem Aufwand das JSON einzulesen. Vielleicht als kleines Beispiel. Hintergrund, wenn wir schon sowas macchen, könnte man nur von einem Punkt einlesen und dann ist das JSON der Datenbestand. Darin muss ich ggf noch das Modul verankern und schon können wir uns daraus die Daten holen und pflegen nur an einer Stelle wie gewollt :-)

Reply:
Ja, ich denke das ist relativ leicht machbar.

Reply:
Ich habe mich mal an unserem bestehenden JSON hinsichtlich Aufbau orientiert:

https://jsoneditoronline.org/?id=b30e595a040e477eaed01f49e038bc40

Wäre das was?

Reply:
Verstehen wir uns Miss?

Ich möchte keine neue Struktur aufbauen. Ich möchte das JSON von uns einlesen als Hash ? um dann via Schleife die passenden Einträge via Modulfilterung abzufragen.

Reply:
Welches JSON möchtest Du denn genau einlesen? Vielleicht habe ich da jetzt was falsch verstanden.
Mit genannter Verlinkung habe ich deinen Test schon mal angepasst.

Klappte leider nicht 1:1 :(

Reply:
Smile,

wie wäre es mit diesem https://github.com/RFD-FHEM/SIGNALduino_TOOL/blob/pre-release/FHEM/lib/SD_Device_ProtocolList.json genau das wo du Wert legst Daten aktuell zu haben für Tests ;)

Reply:
Okay, ich dachte Du meinst ein anderes :) 
Ich denk mal eben schnell drüber nach :)